### PR TITLE
[RFR for API Sources] New Python interfaces to support resumable full refresh

### PR DESCRIPTION
**Pull request by brianjlai**
## What

Primarily updates how the Stream class performs a read. The big change is around how we resolve what the next partition of records to retrieve is. Because resumable full refresh operates under the paradigm of an unbounded set of pages (unlike incremental partitioned time windows), we need to change how we determine the next slice.

The primary changes to the flow at a high level are:

- Consolidates a lot of the branching logic to be agnostic of the incoming catalog's sync mode which incorporated parts of Ella's changes in https://github.com/airbytehq/airbyte/pull/36999 
- Instead of `stream_slices()` being the mechanism for determining the next partition of records to retrieve. I've introduced the concept of a `CheckpointReader` whose type is instantiated based on the Stream's implementation
- The reader interprets and passed state back to the ConnectorStateManager by observing the stream's current state.
- The CDK is now "state aware" meaning that it actually reads stream state instead of just passing it back to the platform like a block box. I'll go into this in more detail below.

## How

Some of the major design changes in the review are:

- The new `CheckpointReader` class which is now the main way a stream determines the next partition of values to read. For incremental this continues to be partitions like time windows. For RFR, this is the next page of records. And for RFR this can be parent records for substreams or a single slice `{}`.
- Deprecating `IncrementalMixin` in favor of `StateMixin` since state is used by RFR streams which are not incremental. This is a better name, but I kept the old one for backwards compatibility reasons
- Changing the default `Stream.stream_slices()` implementation from [None] to [{}]. None is now the indicator to stop iterating over slices
- Adding the `supports_checkpointing` field to streams. It’s needed for two areas. We need to surface this value to the catalog. AND we need this to be overridable because declarative low-code sources delineation for checkpointing differs from python sources.

This PR does not implement the work for a substream that requires resumable full refresh. I have a sketched out interface to see if its possible which it does appear to be. But substream state management for RFR becomes quite convoluted due to the issue I'll go into below.

### Making the CDK and `Stream` class read directly from the connector managed state

This is arguably the most controversial DX and design change compared to before RFR. It has _some_ impact on connector developers. In order for RFR to function using the current `read_records()` method, we need some way of communicating state from the the specific connector implementation back to the CDK. To do so, we expect the developer to manage `Stream.state`. And with RFR we now read the input to decide what to do next. For example:

if state:
`{ "pageNumber": 23 }`
then,
continue syncing.

if state:
`{}`
then,
stop syncing no more pages.

This is a relatively simple example, but it does illustrate that the developer needs to know to emit `{}` to stop RFR paging. In hubspot updating state is done with `self.state = self.next_page_token(response) or {}`. But it feels like not an ideal precedent that a developer needs to have a general awareness of how the checkpoint reader works in order to successfully implement RFR. And this is coupled with the `CheckpointReader` being mostly an internal CDK implementation detail that developers shouldn't need to think of.

As mentioned earlier, substream RFR streams which would need per-partition cursors requires much more careful reading of the state object for specific structure which is indicative that state as an unstructured map is not the right data type.

This is just for connectors that implement the legacy Python CDK, and since state is managed internally by low-code connectors, I think its fair to also say from the 80:20 rule, we are aiming for a more ideal interface for low-code which makes up a majority of our connectors vs legacy Python.

Alternative:

I did also look into incremental vs. RFR streams having different versions of the `read_records()` method. For incremental/full refresh, they would have the normal `read_records() -> iterable`. And for RFR `read_records() -> iterable + updated stream state`. This however felt like a step in the wrong direction because we're moving back into two different runtime flows depending on sync type.

A structured `state` class:

This feels like the more appropriate long term solution. If we have a state interface that handles interpreting state and explicitly communicate what to do with state and what the CDK should emit back to the platform. This is well outside what we can do for the project, but something that I thought about.

## Review guide

1. `checkpoint_reader.py`
2. `core.py`
3. `declarative_stream.py`
4. `abstract_source.py`
5. test files

## User Impact

noop

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:55]: I renamed this mixin because state setter/getter are no longer specific to Incremental streams. But to retain backwards compatibility I left the old `IncrementalMixin` but it just inherits from this mixin

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:103]: Because we call `get_checkpoint()` and the end of the last slice/page and at the end of the sync, we end up emitting the same final state twice. We can potentially insert more fields to track state internally within the reader, but i don't think its worth the hassle

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py
Comment (brianjlai) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:1389]: I left the implementation to get the 2 RFR streams working for Hubspot within the new interfaces to illustrate how to uptake the changes in the CDK.

But before merging I'll remove these from the PR

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:38]: See the comment a few lines above. Ideally we should simply have one checkpoint method that is called at the end of a slice or after syncing all records of a stream that reads `stream.state` and the checkpoint reader doesn't have to manage its own internal state.

However, we retain the deprecated legacy method `get_updated_state()` which manages state by comparing it against emitted records. The legacy behavior forces us to decouple updating state and when to emit state since we otherwise would emit state too frequently after every record.

Given the checkpoint reader is not a public interface or intended for use outside the CDK, we can tolerate the interface not being perfect, but I did want to highlight why I had to do this

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:90]: I've talked about this aspect at length with a few other engineers. I really don't know a better way forward in the immediate right now. Having it be non-obvious that the connector developer be state aware is not a good DX.

The longer term solution would either be a structured state class that handles how to read/write state using clearly defined method instead of a generic map. Or through a cursor class. The problem is it doesn't current exist in the legacy Python CDK

However, we can potentially rationalize that this is okay in the immediate for two reasons:
- This should be less of a problem for low-code because the actual state management code is owned by us and not the connector developer. It also already supports cursors
- We also want to start moving more things to concurrent where we already have a concept of cursors and is worth integrating there as well.

Basically, even if this interface is not a good paradigm shift in the short term, if we're focusing on concurrent + low-code long term, we can accept this. Happy to be challenged on this as well

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:148]: This was a tricky edge case to note. Some CDKs like file-based ones as part of the `self.state` setter will actually manipulate the state that gets set onto cursor. So rather than using the incoming `stream_state` parameter coming in, we should attempt to use the actual value stored on the stream's state. This was caught in tests.

For example for a file based connector:
Incoming State:
```
{
  "history": {
      "b.csv": "2023-06-05T03:54:07.000000Z",
      "c.csv": "2023-06-05T03:54:07.000000Z",
      "d.csv": "2023-06-05T03:54:07.000000Z",
}
```

When first setting state for the stream, we call `self.state` > `self._cursor.set_initial_state(value)`. This in turns sets state to:
```
{
    "history": {
        "b.csv": "2023-06-05T03:54:07.000000Z",
        "c.csv": "2023-06-05T03:54:07.000000Z",
        "d.csv": "2023-06-05T03:54:07.000000Z",
    },
    "_ab_source_file_last_modified": "2023-06-05T03:54:07.000000Z_d.csv",
}
```

So the last modified value is also set just by setting the state for the first time. We should try to avoid using the `stream_state` parameter whenever possible.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:272]: I see 3 paths forward:
1. Accept substreams are not resumable
2. Implement support for RFR and substreams in concurrent
3. Implement a something similar to SubstreamPartitionRouter for non-low code connectors

I believe (2) and (3) would be a similar level of effort so I have a preference for (2).

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:282]: wow

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:351]: I agree. Do you want to keep the comment in code for posterity or can we remove it?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:406]: fun fact: slices is a generator so it won't even be logged properly

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:38]: It's just that now there's an extra interface method that could have otherwise been consolidated. Not a deal breaker but aside from the legacy case, we were always following up the update call w/ the emit state.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:90]: yes good point yep I'll find a place to document this either in code or in airbyte docs

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:108]: Since these checkpoint readers aren't intended to be configured by the developer, I think it would make more sense to document this where I am going to document the state object shape mentioned above. It'll be in a separate airbyte docs page

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:148]: yep I can add a permanent comment about why we need this

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:272]: Yep I am in agreement as well. This will be part of the concurrent RFR prototype, which conveniently the Stripe endpoint we want to implement is a substream

https://github.com/airbytehq/airbyte-internal-issues/issues/7390

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:351]: removing

File: airbyte-cdk/python/airbyte_cdk/sources/streams/core.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/core.py:406]: my god this is awful. I'm just getting rid of this

File: docs/connector-development/cdk-python/resumable-full-refresh-stream.md
Comment (brianjlai) [inline docs/connector-development/cdk-python/resumable-full-refresh-stream.md:1]: @girarda fyi i added a docs for RFR instead of inlining comments for certain methods etc. I still need to proof grammar and clean it up a little but just a heads up since some PR comments called this out

File: airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:38]: can you add a comment explaining that `None` means we stop reading?

Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py:38]: I think you meant this for the `checkpointReader.next()`, since None in that case stops parsing, but I'll also comment here that we don't emit state messages if return is None either.

**Review by girarda — APPROVED** [review]
this is great! approved pending reverting the changes to hubspot

**Comment by brianjlai** [comment]
ran a few local `connectors-ci` runs of hubspot, a few low-code connectors to see that tests passed before the merge

**Pull request merged**
